### PR TITLE
Code maintenance/46647 bump rails to 7.1

### PR DESCRIPTION
**Pull request by dombesz**
https://community.openproject.org/wp/46647

### CODE REVIEW

File: modules/bim/db/migrate/20210521080035_update_xkt_to_version8.rb [hunk 1]
```diff
@@ -31,11 +31,20 @@ class UpdateXktToVersion8 < ActiveRecord::Migration[6.1]
                         'Hospital - Structural (cc-by-sa-3.0 Autodesk Inc.)' => 'hospital_structural',
                         'Hospital - Mechanical (cc-by-sa-3.0 Autodesk Inc.)' => 'hospital_mechanical' }.freeze
 
+  # Note: rails 7.1 breaks the class' ancestor chain, and raises an error, when a class
+  # with an enum definition without a database field is being referenced.
+  # Re-defining the Project class without the enum to avoid the issue.
```
Comment (aaron-contreras) [inline modules/bim/db/migrate/20210521080035_update_xkt_to_version8.rb]:
```suggestion
  # Re-defining the IfcModel class without the enum to avoid the issue.
```

File: modules/bim/db/migrate/20211011204301_add_column_conversion_status_to_ifc_model.rb [hunk 2]
```diff
@@ -27,15 +27,24 @@
 #++
 
 class AddColumnConversionStatusToIfcModel < ActiveRecord::Migration[6.1]
+  # Note: rails 7.1 breaks the class' ancestor chain, and raises an error, when a class
+  # with an enum definition without a database field is being referenced.
+  # Re-defining the Project class without the enum to avoid the issue.
```
Comment (aaron-contreras) [inline modules/bim/db/migrate/20211011204301_add_column_conversion_status_to_ifc_model.rb]: Same here

**Review by aaron-contreras — COMMENTED** [review]
Looks good! I believe the specs are failing because using `Rails.application.credentials` now expects to find these secrets in a `credentials.yml` file instead of a `secrets.yml`.

File: app/models/query/results/group_by.rb [hunk 3]
```diff
@@ -163,7 +163,7 @@ def aliased_group_by_sort_order(alias_name, sortable, columns_hash, order = nil)
     column = case_insensitive_condition(sortable, column, columns_hash)
 
     if order
-      column + " #{order} "
+      column.to_s + " #{order} "
```
Comment (aaron-contreras) [inline app/models/query/results/group_by.rb]: Nice find. 👍

File: modules/team_planner/spec/support/pages/team_planner.rb [hunk 4]
```diff
@@ -343,12 +343,7 @@ def y_center(element)
 
     def wait_for_loaded
       expect(page).to have_css('.op-team-planner--wp-loading-skeleton')
-
-      retry_block do
-        raise "Should not be there" if page.has_selector?('.op-team-planner--wp-loading-skeleton')
-
-        sleep 5
-      end
+      expect(page).to have_no_css('.op-team-planner--wp-loading-skeleton')
```
Comment (aaron-contreras) [inline modules/team_planner/spec/support/pages/team_planner.rb]: Perhaps extend the wait on this? I've seen this retry twice/thrice quite consistently on the CI logs with the previous implementation so it takes at least 5-10 seconds for it to be gone. Maybe:

```suggestion
      expect(page).to have_no_css('.op-team-planner--wp-loading-skeleton', wait: 10)
```

File: app/models/query/results/group_by.rb (hunk 3 above)
Comment (toy) [inline app/models/query/results/group_by.rb]: Will this work:
```suggestion
      "#{column} #{order} "
```
Also wondering if space is needed at the end:
```suggestion
      "#{column} #{order}"
```

File: config/environments/test.rb [hunk 5]
```diff
@@ -26,47 +26,56 @@
...
 
-  # The test environment is used exclusively to run your application's
-  # test suite. You never need to work with it otherwise. Remember that
-  # your test database is "scratch space" for the test suite and is wiped
-  # and recreated between test runs. Don't rely on the data there!
-  #
   # Spring requires to have the classes reloaded. On the CI or when Spring is
   # disabled, it does not need to happen.
-  config.cache_classes = ENV['CI'].present? || ENV['DISABLE_SPRING'].present?
+  config.enable_reloading = !(ENV['CI'].present? || ENV['DISABLE_SPRING'].present?)
```
Comment (toy) [inline config/environments/test.rb]: Will this be simpler to read?
```suggestion
  config.enable_reloading = %w[CI DISABLE_SPRING].none?{ |name| ENV[name].present? }
```

File: Gemfile [hunk 6]
```diff
@@ -39,7 +39,7 @@ gem 'activemodel-serializers-xml', '~> 1.0.1'
 gem 'activerecord-import', '~> 1.5.0'
 gem 'activerecord-session_store', '~> 2.1.0'
 gem 'ox'
-gem 'rails', '~> 7.0', '>= 7.0.3.1'
+gem 'rails', '~> 7.1.2', '>= 7.1.3'
```
Comment (aaron-contreras) [inline Gemfile]: Curious :) Is using this versioning syntax any different than:

```suggestion
gem 'rails', '~> 7.1.3'
```

I understand the current syntax as anything above "7.1.2 and below 7.2" AND "at least 7.1.3" Doesn't the above suggestion say the same a bit more concisely?

File: app/controllers/repositories_controller.rb [hunk 7]
```diff
@@ -46,7 +46,7 @@ class RepositoriesController < ApplicationController
 
   before_action :find_project_by_project_id
   before_action :authorize
-  before_action :find_repository, except: %i[edit update create destroy destroy_info]
+  before_action :find_repository, except: %i[update create destroy destroy_info]
```
Comment (aaron-contreras) [inline app/controllers/repositories_controller.rb:49]: 👍

Comment (aaron-contreras) [inline app/controllers/repositories_controller.rb:49]: ⭐️

File: app/models/capabilities/scopes/default.rb [hunk 8]
```diff
@@ -155,6 +155,11 @@ def default_sql_by_non_member_with_anonymous
           WHERE enabled_modules.project_id IS NOT NULL OR "actions".module IS NULL
         SQL_PART
       end
+
+      def principal_sql
+        RequestStore[:capabilities_principal_sql] ||=
+          Principal.visible.not_builtin.not_locked.to_sql
+      end
```
Comment (aaron-contreras) [inline app/models/capabilities/scopes/default.rb]: Nice :)

File: app/models/members/scopes/not_locked.rb [hunk 9]
```diff
@@ -35,7 +35,7 @@ module NotLocked
       def not_locked
         includes(:principal)
           .references(:principals)
-          .merge(Principal.not_locked, rewhere: true)
+          .merge(Principal.not_locked)
```
Comment (aaron-contreras) [inline app/models/members/scopes/not_locked.rb:38]: What was `rewhere: true` doing before? Couldn't find any good answers on this around

File: app/models/project.rb [hunk 10]
```diff
@@ -149,7 +149,6 @@ class Project < ApplicationRecord
 
   friendly_id :identifier, use: :finders
 
-  include ::Scopes::Scoped
```
Comment (aaron-contreras) [inline app/models/project.rb:152]: Nice catch on the duplicate.

File: app/workers/mails/mailer_job.rb [hunk 11]
```diff
@@ -41,7 +41,7 @@ class Mails::MailerJob < ApplicationJob
   queue_as { ActionMailer::Base.deliver_later_queue_name }
 
   # Retry mailing jobs three times with exponential backoff
-  retry_on StandardError, wait: :exponentially_longer, attempts: 3
+  retry_on StandardError, wait: :polynomially_longer, attempts: 3
```
Comment (aaron-contreras) [inline app/workers/mails/mailer_job.rb]:
```suggestion
  # Retry mailing jobs three times with polynomial backoff
  retry_on StandardError, wait: :exponentially_longer, attempts: 3
  retry_on StandardError, wait: :polynomially_longer, attempts: 3
```

File: config/environments/test.rb [hunk 12]
```diff
@@ -75,18 +85,30 @@
...
-  # Raises error for missing translations
+  # Raises error for missing translations.
   config.i18n.raise_on_missing_translations = true
 
+  # Annotate rendered view with file names.
+  # config.action_view.annotate_rendered_view_with_filenames = true
+
+  # Raise error when a before_action's only/except options reference missing actions
+  config.action_controller.raise_on_missing_callback_actions = true
+
```
Comment (aaron-contreras) [inline config/environments/test.rb:111]: ⭐️

File: config/initializers/filter_parameter_logging.rb [hunk 13]
```diff
@@ -28,5 +28,9 @@
...
 # Be sure to restart your server when you modify this file.
 
-# Configure sensitive parameters which will be filtered from the log file.
-Rails.application.config.filter_parameters += [:password]
+# Configure parameters to be partially matched (e.g. passw matches password) and filtered from the log file.
+# Use this to limit dissemination of sensitive information.
+# See the ActiveSupport::ParameterFilter documentation for supported notations and behaviors.
+Rails.application.config.filter_parameters += %i[
+  passw secret token _key crypt salt certificate otp ssn
+]
```
Comment (aaron-contreras) [inline config/initializers/filter_parameter_logging.rb:36]: Nice

File: lib/open_project/static_routing.rb [hunk 14]
```diff
@@ -37,9 +37,11 @@ module StaticRouting
     # Most importantly it does work for the '#{model}_path|url' helpers, though.
     module UrlHelpers
       extend ActiveSupport::Concern
-      include Rails.application.routes.url_helpers
 
       included do
+        # See https://github.com/rails/rails/pull/50403
+        include Rails.application.routes.url_helpers
+
         def default_url_options
```
Comment (aaron-contreras) [inline lib/open_project/static_routing.rb:45]: Interesting order of operations! Thanks for linking the pull request

File: modules/boards/spec/features/support/board_page.rb [hunk 15]
```diff
@@ -136,9 +136,13 @@ def expect_not_referencable(list_name, work_package)
     # Expect the given titled card in the list name to be present (expect=true) or not (expect=false)
     def expect_card(list_name, card_title, present: true)
       within_list(list_name) do
+        # Wait for the card loading to finish
+        expect(page).to have_no_selector(".loading-indicator--background")
         expect(page).to have_conditional_selector(present,
                                                   '[data-test-selector="op-wp-single-card--content-subject"]',
-                                                  text: card_title)
+                                                  text: card_title,
+                                                  # Wait only 3 seconds for non-presence expectation
```
Comment (aaron-contreras) [inline modules/boards/spec/features/support/board_page.rb]: I think the comment no longer applies.

```suggestion
                                                  # Don't wait on non-presence expectation
```

File: modules/costs/app/controllers/costlog_controller.rb [hunk 16]
```diff
@@ -30,7 +30,6 @@ class CostlogController < ApplicationController
   menu_item :work_packages
   before_action :find_project, :authorize, only: %i[edit new create update destroy]
   before_action :find_associated_objects, only: %i[create update]
-  before_action :find_optional_project, only: %i[report]
```
Comment (aaron-contreras) [inline modules/costs/app/controllers/costlog_controller.rb:33]: 👍

File: modules/dashboards/spec/features/project_status_spec.rb [hunk 17]
```diff
@@ -118,7 +118,7 @@ def add_project_status_widget
         field.set_to('AT RISK')
 
         # The edit field is toggled and the value saved.
-        expect(page).to have_content('AT RISK')
+        expect(page).to have_content('AT RISK', wait: 5)
```
Comment (aaron-contreras) [inline modules/dashboards/spec/features/project_status_spec.rb:121]: This is not a request for change, but something to think about amongst all of us. It would be nice to agree on a non-arbitrary long wait time for content that is AJAX-dependent and use a constant throughout the app like:

```ruby
AJAX_DEPENDENT_WAIT = 15.seconds # or whatever number we decide to try
expect(page).to have_content('AT RISK', wait: AJAX_DEPENDENT_WAIT)
```

File: modules/meeting/app/controllers/meeting_agenda_items_controller.rb [hunk 18]
```diff
@@ -35,7 +35,7 @@ class MeetingAgendaItemsController < ApplicationController
   before_action :set_meeting
   before_action :set_agenda_item_type, only: %i[new create]
   before_action :set_meeting_agenda_item,
-                except: %i[index new cancel_new create author_autocomplete_index]
+                except: %i[new cancel_new create] # author_autocomplete_index
```
Comment (aaron-contreras) [inline modules/meeting/app/controllers/meeting_agenda_items_controller.rb]:
```suggestion
                except: %i[new cancel_new create]
```

File: modules/meeting/spec/models/meeting_acts_as_journalized_spec.rb [hunk 19]
```diff
@@ -78,7 +78,7 @@
 
       it 'has the updated_at of the meeting as the lower bound for validity_period and no upper bound' do
         expect(meeting.last_journal.validity_period)
-          .to eql(meeting.reload.updated_at...Float::INFINITY)
+          .to eql(meeting.reload.updated_at...)
```
Comment (aaron-contreras) [inline modules/meeting/spec/models/meeting_acts_as_journalized_spec.rb:81]: Nice :)

File: modules/reporting/spec/features/support/pages/cost_report_page.rb [hunk 20]
```diff
@@ -87,6 +87,11 @@ def show_loading_indicator(present: true)
       end
     end
 
+    def wait_for_page_to_reload
+      show_loading_indicator
+      show_loading_indicator(present: false)
+    end
+
```
Comment (aaron-contreras) [inline modules/reporting/spec/features/support/pages/cost_report_page.rb:94]: That's clever!

File: spec/helpers/settings_helper_spec.rb [hunk 21]
```diff
@@ -36,39 +36,44 @@
...
   before do
     allow(Setting)
-      .to receive(:field_writable?)
+      .to receive(:host_name_writable?)
             .and_return true
+
+    allow(I18n).to receive(:t).and_call_original
+    allow(I18n).to receive(:t).with('setting_field_a').and_return('Field A')
+    allow(I18n).to receive(:t).with('setting_field_b').and_return('Field B')
+    allow(I18n).to receive(:t).with('setting_field').and_return('Field')
```
Comment (aaron-contreras) [inline spec/helpers/settings_helper_spec.rb:45]: 👍

File: spec/lib/api/v3/queries/schemas/query_schema_representer_spec.rb [hunk 22]
```diff
@@ -357,7 +357,11 @@
              Queries::WorkPackages::Columns::PropertyColumn.new(:bogus2),
              Queries::WorkPackages::Columns::PropertyColumn.new(:bogus3),
              Queries::WorkPackages::Columns::RelationToTypeColumn.new(type),
-             Queries::WorkPackages::Columns::RelationOfTypeColumn.new(name: :label_relates_to, sym: :relation1)]
+             Queries::WorkPackages::Columns::RelationOfTypeColumn.new(
+               name: :label_relates_to,
+               sym: :relation1,
+               sym_name: :label_relates_to
+             )]
```
Comment (aaron-contreras) [inline spec/lib/api/v3/queries/schemas/query_schema_representer_spec.rb:364]: This was quite a bit of detective work! Nice

File: spec/models/work_package/work_package_acts_as_journalized_spec.rb [hunk 23]
```diff
@@ -379,9 +379,8 @@
 
         it 'does not set an upper bound to the already existing journal' do
           work_package.save
-
-          expect(work_package.last_journal.validity_period.last)
-            .to eql(Float::INFINITY)
+          expect(work_package.last_journal.validity_period.end)
+            .to be_nil
```
Comment (aaron-contreras) [inline spec/models/work_package/work_package_acts_as_journalized_spec.rb:383]: Your grasp on Journals is way further along than mine. All I'm wondering is if this has any behavioral implication

**Review by aaron-contreras — CHANGES_REQUESTED** [review]
Hell of a job! 🎸

I mostly just left some cleanup to-dos and questions

```
                     `. ___
                    __,' __`.                _..----....____
        __...--.'``;.   ,.   ;``--..__     .'    ,-._    _.-'
  _..-''-------'   `'   `'   `'     O ``-''._   (,;') _,'
,'________________                          \`-._`-','
 `._              ```````````------...___   '-.._'-:
    ```--.._      ,.                     ````--...__\-.
            `.--. `-`                       ____    |  |`
              `. `.                       ,'`````.  ;  ;`
                `._`.        __________   `.      \'__/`
                   `-:._____/______/___/____`.     \  `
                               |       `._    `.    \
                               `._________`-.   `.   `.___
                                             SSt  `------'`
```

File: app/models/members/scopes/not_locked.rb (hunk 9 above)
Comment (dombesz) [reply app/models/members/scopes/not_locked.rb:38]: This post explains it https://www.fastruby.io/blog/rails/upgrades/rails-merge-deprecation.html .
Using `.merge(..., rewhere: true)` makes sure that field conditions on the original query are overwritten by the field conditions on the merge query.

For example doing a merge with having `published: true` on the base query and having `published: false` on the merge query: 
 - Without using `rewhere` the query would be  "published = TRUE AND published = FALSE".
 - With `rewhere` the query is "published = FALSE".
 
Now in rails 7.1, the latter behaviour became the default, so using rewhere is deprecated.

File: spec/lib/api/v3/queries/schemas/query_schema_representer_spec.rb (hunk 22 above)
Comment (dombesz) [reply spec/lib/api/v3/queries/schemas/query_schema_representer_spec.rb:364]: Yep, thanks for helping in the process!🙌

File: Gemfile (hunk 6 above)
Comment (dombesz) [reply Gemfile]: Yes, it is the same. I updated the rule so we can install any minor release inside 7.1.x that is above 7.1.3.

File: spec/models/work_package/work_package_acts_as_journalized_spec.rb (hunk 23 above)
Comment (dombesz) [reply spec/models/work_package/work_package_acts_as_journalized_spec.rb:383]: We have a good coverage for testing the aggregation periods so I'm confident that it doesn't change the behaviour. Additionally I verified the database entries and they are also correct.

File: modules/dashboards/spec/features/project_status_spec.rb (hunk 17 above)
Comment (dombesz) [reply modules/dashboards/spec/features/project_status_spec.rb:121]: Yeah, this is definitely something to think about. Right now we have a lot of arbitrary waiting times, depending on certain test cases.

**Comment by dombesz** [comment]
Thanks for the review @aaron-contreras and @toy 🙌. I believe I addressed your comments, please have a look.

File: config/application.rb [hunk 24]
```diff
@@ -114,7 +114,6 @@ class Application < Rails::Application
     config.middleware.use Rack::TempfileReaper
 
     # Custom directories with classes and modules you want to be autoloadable.
-    config.enable_dependency_loading = true
     config.paths.add Rails.root.join('lib').to_s, eager_load: true
```
Comment (toy) [inline config/application.rb]: Is this still needed with addition of `autoload_lib`?

File: app/controllers/application_controller.rb [hunk 25]
```diff
@@ -574,7 +574,7 @@ def pick_layout(*args)
   end
 
   def default_breadcrumb
-    label = "label_#{self.class.name.gsub('Controller', '').underscore.singularize}"
+    label = "label_#{self.class.name.demodulize.gsub('Controller', '').underscore.singularize}"
```
Comment (toy) [inline app/controllers/application_controller.rb]: I think `controller_name` is available here, so this can be replaced with:
```suggestion
    label = "label_#{controller_name.singularize}"
```

File: app/models/capabilities/scopes/default.rb [hunk 26]
```diff
@@ -155,6 +155,11 @@ def default_sql_by_non_member_with_anonymous
           WHERE enabled_modules.project_id IS NOT NULL OR "actions".module IS NULL
         SQL_PART
       end
+
+      def principal_sql
+        RequestStore[:capabilities_principal_sql] ||=
```
Comment (toy) [inline app/models/capabilities/scopes/default.rb]: There are many possible ways of using `RequestStore`, but `RequestStore.fetch` seems to be used more often in the codebase compared with `RequestStore[]`

**Review by aaron-contreras — APPROVED** [review]
Excellent! ⭐️🤌

File: Gemfile [hunk 27]
```diff
@@ -163,7 +163,7 @@ gem 'matrix', '~> 0.4.2'
 
 gem 'meta-tags', '~> 2.20.0'
 
-gem "paper_trail", "~> 12.3"
+gem "paper_trail", "~> 15.1.0"
```
Comment (toy) [inline Gemfile]:
```suggestion
gem 'paper_trail', '~> 15.1.0'
```

File: app/models/token/hashed_token.rb [hunk 28]
```diff
@@ -46,7 +46,7 @@ def self.hash_function(input)
       # Use a fixed salt for hashing token values.
       # We still want to be able to index the hash value for fast lookups,
       # so we need to determine the hash without knowing the associated user (and thus its salt) first.
-      Digest::SHA256.hexdigest(input + Rails.application.secrets.fetch(:secret_key_base))
```
Comment (toy) [inline app/models/token/hashed_token.rb:49]: Was fetch ensuring presence of value which will not be done anymore?

File: app/models/user.rb [hunk 29]
```diff
@@ -500,7 +500,7 @@ def self.execute_as(user, &)
   #   - OmniAuth
   #   - LDAP
   def missing_authentication_method?
-    identity_url.nil? && passwords.empty? && auth_source.nil?
+    identity_url.nil? && passwords.empty? && ldap_auth_source_id.nil?
```
Comment (toy) [inline app/models/user.rb:510]: Why not `!ldap_auth_source_id?`?

File: bin/setup [hunk 30]
```diff
@@ -1,29 +1,33 @@
...
+APP_ROOT = File.expand_path('..', __dir__)
 
-Dir.chdir APP_ROOT do
-  # This script is a starting point to setup your application.
-  # Add necessary setup steps to this file:
+def system!(*args)
+  system(*args, exception: true)
+end
+
+FileUtils.chdir APP_ROOT do
```
Comment (toy) [inline bin/setup]: Unless this is generated there is `Dir.chdir` and there is no need to require fileutils then

File: app/models/queries/work_packages/columns/relation_of_type_column.rb [hunk 31]
```diff
@@ -28,9 +28,8 @@
 
 class Queries::WorkPackages::Columns::RelationOfTypeColumn < Queries::WorkPackages::Columns::RelationColumn
   def initialize(type)
-    super
-
     self.type = type
+    super(name)
```
Comment (toy) [inline app/models/queries/work_packages/columns/relation_of_type_column.rb:32]: Is it an unrelated bug fix?

File: spec/requests/api/v3/user/create_user_resource_spec.rb [hunk 32]
```diff
@@ -117,7 +117,7 @@ def send_request
           expect(last_response.status).to eq(422)
 
           expect(last_response.body)
-            .to be_json_eql('authSource'.to_json)
+            .to be_json_eql('ldapAuthSource'.to_json)
```
Comment (toy) [inline spec/requests/api/v3/user/create_user_resource_spec.rb]: Can this change be important?

File: app/models/queries/work_packages/columns/relation_of_type_column.rb (hunk 31 above)
Comment (aaron-contreras) [reply app/models/queries/work_packages/columns/relation_of_type_column.rb:32]: It relates to https://github.com/opf/openproject/pull/14440#discussion_r1471816323

File: app/models/token/hashed_token.rb (hunk 28 above)
Comment (aaron-contreras) [reply app/models/token/hashed_token.rb:49]: I guess it does, but on the same note, doesn't Rails complain either way if this is not present when running with `RAILS_ENV=production`? Or was is it the master key that it complains about?

Comment (toy) [reply app/models/token/hashed_token.rb:49]: True, I get ``ArgumentError: Missing `secret_key_base` for 'production' environment, set this string with `bin/rails credentials:edit` (ArgumentError)``

File: app/models/user.rb (hunk 29 above)
Comment (dombesz) [reply app/models/user.rb:510]: I just followed the convention in the method, but I also think it's more descriptive to call `.nil?`

File: spec/requests/api/v3/user/create_user_resource_spec.rb (hunk 32 above)
Comment (dombesz) [reply spec/requests/api/v3/user/create_user_resource_spec.rb]: Yep it is important, is has slipped through my radar. Thanks! 🙌

File: app/models/user.rb (hunk 29 above)
Comment (toy) [reply app/models/user.rb:510]: There are multiple variants in this file, using `ldap_auth_source` and `ldap_auth_source_id`, with `blank?`, `present?` or just checking truthiness, so probably better to unify separately from this PR

File: app/models/token/hashed_token.rb (hunk 28 above)
Comment (dombesz) [reply app/models/token/hashed_token.rb:49]: Yes, calling `Rails.application.secret_key_base` also validates the presence of the `secret_key_base` and raises the error you mentioned.
```
[1] pry(main)> Rails.application.secret_key_base

From: /.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/railties-7.1.3/lib/rails/application.rb:654 Rails::Application#validate_secret_key_base:

    652: def validate_secret_key_base(secret_key_base)
    653:   binding.pry
 => 654:   if secret_key_base.is_a?(String) && secret_key_base.present?
    655:     secret_key_base
    656:   elsif secret_key_base
    657:     raise ArgumentError, "`secret_key_base` for #{Rails.env} environment must be a type of String`"
    658:   else
    659:     raise ArgumentError, "Missing `secret_key_base` for '#{Rails.env}' environment, set this string with `bin/rails credentials:edit`"
    660:   end
    661: end
```

**Review by toy — APPROVED** [review]
Looks good, as far as I can tell

**Comment by aaron-contreras** [comment]
Awesome job my friend! @dombesz
